### ISSUE TITLE
"segment cannot be found" in lambda function

### ISSUE THREAD

**Issue by aleist** [aws/aws-xray-sdk-go#50]
I'm new to X-Ray and am trying to integrate it into a Lambda function called from API Gateway as a Lambda proxy integration request. My code follows this example:

```go
xray.Configure(xray.Config{LogLevel: "trace"})
sess := session.Must(session.NewSession())
dynamo := dynamodb.New(sess)
xray.AWS(dynamo.Client)
````

This results in the following error being logged `Error Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 'dynamodb': segment cannot be found.`, which is followed by a panic `runtime error: invalid memory address or nil pointer dereference` in the xray SDK when the first DynamoDB operation executes.

I can "fix" this by creating a new segment with `xray.BeginSegment`, which puts a new service node in the graph.

I have tried this both with "active tracing" enabled and disabled in the Lambda config. When enabled, I get 2 more nodes for the Lambda service and function in the graph, but they're not connected to the new segment I create in the function as described above.

I am not sure if I'm doing something wrong (in which case the example would seem wrong too) or if this is a bug.

It also worries me that this results in a panic that propagates to my Lambda, because even if I fix this particular problem, I don't like the idea of a library that just instruments my code potentially causing a fatal error of the lambda.

**Comment by luluzhao** [comment]
Hi @aleist, thank you for contacting us regarding your issue and I am sorry for the inconvenience. When using X-Ray SDK inside a lambda function, lambda will create a segment and X-Ray SDK will create corresponding subsegment for customer's application, which means you don't have to call `xray.BeginSegment` inside a lambda function. Here I attach an example how to use X-Ray SDK inside a lambda function.

```go
func HandleRequest(ctx context.Context, name string) (string, error) {
    xray.Configure(xray.Config{LogLevel: "trace",})
    sess := session.Must(session.NewSession())
    dynamo := dynamodb.New(sess)
    xray.AWS(dynamo.Client)
    input := &dynamodb.PutItemInput{
        Item: map[string]*dynamodb.AttributeValue{
            "a": {
                S: aws.String("xray"),
            },
        },
        TableName: aws.String("xray"),
    }

    _, err := dynamo.PutItemWithContext(ctx, input)
    if err != nil {
        return name, err
    }

    return fmt.Sprintf("Hello %s!", name), nil
}
````
My assumption is you didn't call AWS API `WithContext` method so that you received `segment cannot be found` exception since we inject segment information into `context` and use it to keep track of the latest segment information. Please update the API method and you won't be able to see the error.

For the panic issue, we failed to capture the nil pointer in our code. We will start working on this bug and release the fix ASAP. Thank you for your patience. Feel free to contact me if you have any questions.

**Comment by aleist** [comment]
Thank you for your reply @luluzhao. What you write confirms my understanding of the way X-Ray functions. I am using `BatchGetItemPagesWithContext` instead of `PutItemWithContext`, but the context is passed along as required.

In case this helps, here is the part of the nil pointer dereference panic relating to the AWS or xray SDKs:

```
<root>/vendor/github.com/aws/aws-xray-sdk-go/xray.glob..func1
<root>/vendor/github.com/aws/aws-xray-sdk-go/xray/aws.go:53
<root>/vendor/github.com/aws/aws-sdk-go/aws/request.(*HandlerList).Run
<root>/vendor/github.com/aws/aws-sdk-go/aws/request/handlers.go:195
<root>/vendor/github.com/aws/aws-sdk-go/aws/request.(*Request).Build
<root>/vendor/github.com/aws/aws-sdk-go/aws/request/request.go:355
<root>/vendor/github.com/aws/aws-sdk-go/aws/request.(*Request).Sign
<root>/vendor/github.com/aws/aws-sdk-go/aws/request/request.go:376
<root>/vendor/github.com/aws/aws-sdk-go/aws/request.(*Request).Send
<root>/vendor/github.com/aws/aws-sdk-go/aws/request/request.go:483
<root>/vendor/github.com/aws/aws-sdk-go/aws/request.(*Pagination).Next
<root>/vendor/github.com/aws/aws-sdk-go/aws/request/request_pagination.go:93
<root>/vendor/github.com/aws/aws-sdk-go/service/dynamodb.(*DynamoDB).BatchGetItemPagesWithContext
<root>/vendor/github.com/aws/aws-sdk-go/service/dynamodb/api.go:205
```

**Comment by luluzhao** [comment]
@aleist, thank you so much for attaching error log message to help us debug. Yes, the nil pointer will happen at `BeginSubsegment` method. Are you still been blocked by the nil pointer issue?

**Comment by luluzhao** [comment]
@aleist , thank you for your information. I will also try to reproduce your issue on my side. Could you tell me which X-Ray Go SDK version are you using?

**Comment by aleist** [comment]
@luluzhao Thank you. I'm using the following SDK versions:

github.com/aws/aws-xray-sdk-go @ v0.9.4
github.com/aws/aws-sdk-go @ v1.13.30
github.com/aws/aws-lambda-go @ v1.1.0

**Comment by luluzhao** [comment]
@aleist, sorry for the late reply. During the debug process, I just realized Lambda support starting from version `v1.0.0-rc.1` and you are using `v0.9.4`. So please update the X-Ray Go SDK to `v1.0.0-rc.2` since I found a bug in our latest release and fix is under code review. I am sorry for the inconvenience.

**Comment by aleist** [comment]
@luluzhao ah I see, my apologies for using an outdated version. I hadn't realised that there are release candidates for v1.0 available, because `dep ensure` had selected v0.9.4 instead of the RC and even `dep status` doesn't show them.

It does indeed work with rc2. Thank you very much for your help!

**Comment by aleist** [comment]
@luluzhao I'm afraid I just ran into the same issue again while trying to add xray to another lambda that uses not only DynamoDB, but also SES and S3 clients. It works fine when I only pass the DynamoDB and SES clients to `xray.AWS`, but when I do the same for the S3 client, I once again get

```
Suppressing AWS X-Ray context missing panic: failed to begin subsegment named 's3': segment cannot be found.
```

Followed by a nil pointer dereference panic when the client is used.

**Comment by luluzhao** [comment]
Hi @aleist, the usage for all aws client should be the same. S3 is not an exception for using. There must be some issues happened behind. Could you provide more information about it like code snippet?

**Comment by aleist** [comment]
@luluzhao In this lambda I only use the S3 client to presign a GET request as follows:

```
func PresignGetObject(client *s3.S3, bucket, key string, expire time.Duration) (string, error) {
	req, _ := client.GetObjectRequest(&s3.GetObjectInput{
		Bucket: aws.String(bucket),
		Key:    aws.String(key),
	})
	return req.Presign(expire)
}
```

I don't really need to trace this with xray in this particular case, as it doesn't send any S3 service requests over the network, but I'd still expect it to work with the client "xray enabled", as I may need to add an actual S3 request that does go over the network and which I would like to trace later on.

**Comment by luluzhao** [comment]
@aleist, thank you so much for providing the details. Currently our X-Ray SDK doesn't support tracing presign a request. We only trace a really request that send over through the network. In the meantime, we should do a nil check instead of panic if it's a presign request. I will add this into our backlog.

**Comment by jabalsad** [comment]
Hi @luluzhao yes to clarify, this is also inside a lambda function. I've tried `v1.0.0-rc.1`, `v1.0.0-rc.2` and `v1.0.0-rc.3` but I get the segment error inside the lambda.

**Comment by luluzhao** [comment]
@jabalsad, thank you for your clarification. Based on the code you attached, why not use `client` to call `GetItemWithContext`? I don't know what is `d.client`. What is more, could you follow the example I attached in previous reply and refactor the code or you can just try with my example to see whether it is work or not.

**Comment by luluzhao** [comment]
@jabalsad, I am sorry I didn't refresh the page before I replied so that I didn't see your attached example. In `GetSomething` method, it requests to pass a `context` parameter. Have you passed the Lambda function `context` into it? Like I said in the previous reply, Lambda will create and inject the segment into `context` so that our X-Ray Go SDK can use it to create corresponding subsegment.

**Comment by jabalsad** [comment]
`d.client` is a DynamoDB client:

```
func New(table string) *ddb {
	client := dynamodb.New(awsconfig.Session)
	xray.Configure(xray.Config{LogLevel: "trace"})
	xray.AWS(client.Client)
	return &ddb{
		client:     client,
		table:      table,
	}
}
```

The `ctx` is one that is supplied by [aws-lambda-go](https://github.com/aws/aws-lambda-go) such as in [this example](https://github.com/aws/aws-lambda-go/blob/master/events/README_ApiGatewayEvent.md)

**Comment by luluzhao** [comment]
Hi @jabalsad, based on the code snippet you attached, I failed to reproduce the issue you had because it worked on my local box. Would you mind providing all the relevant code to help us debug the issue you had? Thank you.

**Comment by jabalsad** [comment]
@luluzhao I'll have to create a test package to share the code with you which may take a while. Can you provide the code you wrote to reproduce the issue in the meantime? I may be able to spot a difference between your version and mine.

**Comment by luluzhao** [comment]
@jabalsad , here I attached the example I tried.

```go
type ddb struct {
    client *dynamodb.DynamoDB
    table string
}

func New(table string) *ddb {
    client := dynamodb.New(session.Must(session.NewSession()))
    xray.Configure(xray.Config{LogLevel: "trace"})
    xray.AWS(client.Client)
    return &ddb{
        client:     client,
        table:      table,
    }
}

func (d *ddb) GetSomething(ctx context.Context, id string) (string, error) {
    input := &dynamodb.GetItemInput{
        TableName: aws.String(d.table),
        Key:       map[string]*dynamodb.AttributeValue{"xray": {S: aws.String(id)}},
    }

    _, err := d.client.GetItemWithContext(ctx, input)
    if err != nil {
        return "error", err
    }

    return fmt.Sprintf("Hello %s!", id), nil
}


func HandleRequest5(ctx context.Context, id string) (string, error) {
    d := New("xray")
    d.GetSomething(ctx, id)

    return fmt.Sprintf("Hello %s!", id), nil
}

func main() {
    lambda.Start(HandleRequest5)
}
````

**Comment by rengers** [comment]
It turns out we were using a call to `d.client.GetItem(input)` I.e the call that doesn't take in a context.

I would still call this a bug though, since it causes a panic, instead of just failing to add a subsegment.

https://github.com/aws/aws-xray-sdk-go/blob/master/xray/aws.go#L56
```
ctx, opseg := BeginSubsegment(r.HTTPRequest.Context(), r.ClientInfo.ServiceName)
opseg.Namespace = "aws"
```
BeginSubsegment can return `nil, nil`, which I suspect is what is happening here.

1. I don't think we should be returning an `nil` context (This isn't the bug, but is probably not a good idea)
2. We should check `opseg` before trying to use it.

**Comment by dvelop-swil** [comment]
> @aleist, thank you so much for providing the details. Currently our X-Ray SDK doesn't support tracing presign a request. We only trace a really request that send over through the network. In the meantime, we should do a nil check instead of panic if it's a presign request. I will add this into our backlog.

@luluzhao: Hi i didn´t see any fix regarding this issue and still runing in to it on rc.14. Actually I have the same call to get a presign URI as @aleist, with a xray enabled client. In my case the same client is used to also make other calls with context like GetObjectWithContext. So disable xray isn´t really an option here. Did i miss something here or does it or will it not work in this scenario.

**Comment by ashish-lahane-compare-the-market** [comment]
Even I'm facing exactly the same issue. 

go version go1.13.1 linux/amd64
github.com/aws/aws-xray-sdk-go v1.0.0-rc.14

**Comment by ashish-lahane-compare-the-market** [comment]
> Even I'm facing exactly the same issue.
> 
> go version go1.13.1 linux/amd64
> github.com/aws/aws-xray-sdk-go v1.0.0-rc.14

Actually, today it's working fine. My machine rebooted and just clean-rebuilt everything and it worked. Yesterday, I think, the old and incompatible version of aws-xray-sdk-go must have been lurking around somehow. 

Apologies for the trouble.

**Comment by willarmiros** [comment]
Hi @dvelop-swil,
Are you having the same issue of getting a nil pointer panic? There was this [change](https://github.com/aws/aws-xray-sdk-go/commit/9a40d6bf6ff4417cc6558826a72913463800350a) to prevent panics from `nil` segments in `BeginSubegment` for AWS requests, but it doesn't check the context for being `nil`.

**Comment by dvelop-swil** [comment]
Hi @willarmiros,

i get the following error Message. It just says that the segment could not be found, doesn't seem to be a nil error:

`failed to begin subsegment named 's3': segment cannot be found.: string
[
{
    "path": "github.com/aws/aws-lambda-go@v1.13.2/lambda/function.go",
    "line": 31,
    "label": "(*Function).Invoke.func1"
}
,
{
    "path": "runtime/panic.go",
    "line": 522,
    "label": "gopanic"
}
,
{
    "path": "github.com/aws/aws-xray-sdk-go@v1.0.0-rc.14/strategy/ctxmissing/default_context_missing.go",
    "line": 45,
    "label": "(*DefaultRuntimeErrorStrategy).ContextMissing"
}
,
{
    "path": "github.com/aws/aws-xray-sdk-go@v1.0.0-rc.14/xray/segment.go",
    "line": 178,
    "label": "BeginSubsegment"
}
,
{
    "path": "github.com/aws/aws-xray-sdk-go@v1.0.0-rc.14/xray/aws.go",
    "line": 60,
    "label": "glob..func1"
}
,
{
    "path": "github.com/aws/aws-sdk-go@v1.25.23/aws/request/handlers.go",
    "line": 261,
    "label": "(*HandlerList).Run"
}
,
{
    "path": "github.com/aws/aws-sdk-go@v1.25.23/aws/request/request.go",
    "line": 382,
    "label": "(*Request).Build"
}
,
{
    "path": "github.com/aws/aws-sdk-go@v1.25.23/aws/request/request.go",
    "line": 403,
    "label": "(*Request).Sign"
}
,
{
    "path": "github.com/aws/aws-sdk-go@v1.25.23/aws/request/request.go",
    "line": 350,
    "label": "getPresignedURL"
}
,
{
    "path": "github.com/aws/aws-sdk-go@v1.25.23/aws/request/request.go",
    "line": 305,
    "label": "(*Request).Presign"
}`

**Comment by bhautikpip** [comment]
Hi @dvelop-swil,

As per my understanding we still do not have support for presign request. I will add this to our backlog.

**Comment by bhautikpip** [comment]
Hi @dvelop-swil ,

I understood your problem. Follow up question from my side.

1. It looks difficult to reproduce this issue from my end. Would you mind sending me more details about repro steps or may be a code snippet you are using for your use case or any code which can potentially reproduce context missing panic ? 

2. Also, would you mind trying out setting this environment variables as key value pair on your lambda console to test out if it works for your case. There will be environment variable section on your lambda console where you should set key : AWS_XRAY_CONTEXT_MISSING and value: LOG_ERROR. As per my understanding if you set this environment variable then your lambda function should not crash and it should log the error and continue executing your function code.

**Comment by dvelop-swil** [comment]
Hi @bhautikpip 

1. I don't have a full working snippet but here are the steps I'am doing:

`
session, err := session.NewSession(&aws.Config{
		Region: aws.String(endpoints.EuCentral1RegionID)},
)
if err == nil {
	xray.AWSSession(session)
}

s3Client := s3.New(session)

req, _ := s3Client.GetObjectRequest(&s3.PutObjectInput{
	Bucket: aws.String(bucketName),
	Key:    aws.String(objectKey),
})
return req.Presign(5 * time.Minute)
`
On the presign Call the panic occures.

2. Setting the environment variable doesn't suppress the panic. The panic says: "failed to begin subsegment named 's3': segment cannot be found." perhaps there is another environment variable for this?

**Comment by bhautikpip** [comment]
Hi @dvelop-swil,

I tried the attached code. I was not able to see any panic during the presign call. The lambda function executed successfully and x-ray tracing was enabled. 

	sess, err := session.NewSession(&aws.Config{
        Region: aws.String("us-west-2")},
    )

    svc := s3.New(sess)


	req, _ := svc.GetObjectRequest(&s3.GetObjectInput{
        Bucket: aws.String("myBucket"),
        Key:    aws.String("myKey"),
    })
	
	urlStr, err := req.Presign(15 * time.Minute)

	if err != nil {
        log.Println("Failed to sign request", err)
    }

    log.Println("The URL is", urlStr)

**Comment by dvelop-swil** [comment]
Hi @bhautikpip,
in your case you define your own Segment with a Context thats not what iam doing. I activate xray on the session and so for every client has xray activated, your s3Client **svc** hasn't xray activated there for no panic occures.
Just activate xray on the session befor creating the client or on the client it self. e.g.:

```
import (
     "github.com/aws/aws-lambda-go/lambda"
     "github.com/aws/aws-sdk-go/aws"
    "github.com/aws/aws-sdk-go/aws/session"
    "github.com/aws/aws-sdk-go/service/s3"
    "github.com/aws/aws-xray-sdk-go/xray"
    "log"
    "time"
    "context"

)

func PresignGetObject(ctx context.Context) {
    
    ctx, subSeg := xray.BeginSubsegment(ctx, "try_subsegment")

	sess, err := session.NewSession(&aws.Config{
        Region: aws.String("us-west-2")},
    )
   if err == nil {
	xray.AWSSession(sess)
  }

    svc := s3.New(sess)


	req, _ := svc.GetObjectRequest(&s3.GetObjectInput{
        Bucket: aws.String("myBucket"),
        Key:    aws.String("myKey"),
    })
	
	urlStr, err := req.Presign(15 * time.Minute)

	if err != nil {
        log.Println("Failed to sign request", err)
    }

    log.Println("The URL is", urlStr)

    subSeg.Close(nil)

}

func Handler(ctx context.Context) {
	PresignGetObject(ctx)
}

func main() {
	lambda.Start(Handler)
}
```

**Comment by bhautikpip** [comment]
Hi @dvelop-swil,

Thanks for pointing this out. I have attached work around solution which will suppress the panic and will continue execute your function. I will submit a PR for this issue so that you can also set environment variable on lambda console as permanent solution to this issue. currently aws-xray-go-sdk does not support setting up AWS_XRAY_CONTEXT_MISSING environment variable on lambda. 

```
import (
	"context"
	"github.com/aws/aws-lambda-go/lambda"
	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/s3"
	"github.com/aws/aws-xray-sdk-go/xray"
	"log"
	"os"
	"time"
)

func PresignGetObject(ctx context.Context) {
	//ctx context.context

	ctx, subSeg := xray.BeginSubsegment(ctx, "try_subsegment")

	sess, err := session.NewSession(&aws.Config{
		Region: aws.String("us-west-2")},
	)

	if err == nil {
			xray.AWSSession(sess)
	}

	svc := s3.New(sess)

	req, _ := svc.GetObjectRequest(&s3.GetObjectInput{
		Bucket: aws.String("myBucket"),
		Key:    aws.String("myKey"),
	})

	urlStr, err := req.Presign(15 * time.Minute)

	if err != nil {
		log.Println("Failed to sign request", err)
	}

	log.Println("The URL is", urlStr)

	subSeg.Close(nil)

}

func Handler(ctx context.Context) {

	PresignGetObject(ctx)
}

func main() {

	os.Setenv("AWS_XRAY_CONTEXT_MISSING", "LOG_ERROR")
	xray.Configure(xray.Config{
		LogLevel: "",
		LogFormat: "",
	})
	lambda.Start(Handler)
}

```

**Comment by dvelop-swil** [comment]
Hi @bhautikpip,

setting the environment variable directly in the lambda worked. Thanks for the work around!

**Comment by bhautikpip** [comment]
Hi @dvelop-swil,

I merged the fix for it (https://github.com/aws/aws-xray-sdk-go/pull/161). Now you do not have to use work around and you can simply set the environment variable AWS_XRAY_CONTEXT_MISSING as key and LOG_ERROR/RUNTIME_ERROR as value on lambda console. One important thing to note is if you do not set anything then by default  value is LOG_ERROR. I am closing this issue. Feel free to open new issue if this problem still persists.
